core/src/consensus.rs

```diff
@@ -404,9 +486,26 @@ impl Tower {
         total_stake: u64,
         epoch_vote_accounts: &HashMap<Pubkey, (u64, Account)>,
     ) -> SwitchForkDecision {
+        let mut stray_restored_ancestors = HashSet::default();
+
         self.last_voted_slot()
             .map(|last_voted_slot| {
-                let last_vote_ancestors = ancestors.get(&last_voted_slot).unwrap();
+                let last_vote_ancestors = if !self.use_stray_restored_ancestors(ancestors, last_voted_slot) {
+                    ancestors.get(&last_voted_slot).unwrap()
+                } else {
+                    // Use stray restored ancestors because we couldn't derive them from
+                    // given ancestors (=bank_forks)
+                    // Also, can't just return empty ancestors because we should exclude
```

Comment:
From https://github.com/solana-labs/solana/pull/9902#issuecomment-627875209,

> with `unwrap_or_default()`. With that change, these checks should all be skipped: 

I originally returned here `HashSet::default()` but it turned out this should return a set containing `G` (in the original term; in this pr, this is called `stray`).

The corresponding test assertion is https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR2404-R2405.

Reply:
hmmm so this is interesting. 

So first we note that the purpose of `last_vote_ancestors` is to make sure we don't count lockouts on ancestors of the last vote, i.e. to fulfill the check `!llast_vote_ancestors.contains(lockout_interval_start)`, where `lockout_interval_start` > root (checked at https://github.com/solana-labs/solana/blob/master/core/src/consensus.rs#L487)

Now in cases where blockstore is not corrupted, returning the default `HashSet::default()` here should be ok, because on restart from blockstore. To see this we consider the following two cases:

1) If the snapshot root is <= last voted slot, then: 
a) your local root must be a descendant/ancestor of the snapshot root (otherwise we reject that snapshot b/c no `anchor` slots will be found)
b) all your voted slots are descendants of the local root.

Then all the ancestors between the snapshot root and `last voted slot` are present in blockstore at startup and should be replayed at startup. This means `ancestors.get(&last_voted_slot).unwrap()` will return those ancestors. Now I think this test `test_switch_threshold_across_tower_reload()` is failing with `unwrap_or_default()` because when you simulate the reload, the `forks` tree is missing the votes in tower. I suspect if you change the forks tree here: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR2437 to include the votes like so:

```
let forks = tr(0)
            / (tr(1)
                / (tr(2)
                    / tr(10)
                    / (tr(43)
                        / (tr(44)
                            // Minor fork 2
                            / (tr(45) / (tr(46) / (tr(47) / (tr(48) / (tr(49) / (tr(222)))))))
                            / (tr(110) / tr(111))))));
```

things should work. 

Now this does raise the question what happens if your blockstore is corrupted and you can't replay those voted slots on startup. I think the problem with the current approach is that the votes on the stack do not fully capture the ancestor history. For instance, if your voting history was `1, 2, 3, 6`, the stack will only show `1, 2, 6`. 

In such situations (I don't think we have to do these in this PR), I believe the correct thing to do would be a) Return `FailedSwitch` if `switch_slot < last_voted_slot`, and then if `switch_slot > last_voted_slot`, only count `lockout_interval_start > last_voted_slot` so you're sure `lockout_interval_start` is not an ancestor of `last_voted_slot`

2) If the snapshot root is > last voted slot, then because `lockout_interval_start > root` this means that that  `lockout_interval_slot` is not an ancestor of `last_voted_slot` means any  using `HashSet::default` will return the correct result for `!llast_vote_ancestors.contains(lockout_interval_start)`.

Reply:
>     1. otherwise we reject that snapshot b/c no `anchor` slots will be found)

Oops, we weren't rejecting that snapshot as a form of `TowerError` even if our local tower has been rooted. Fixed that: https://github.com/solana-labs/solana/pull/10718/commits/929ca6d859b697d54753e1e4b85c04c85d7822a0#r470455410

Reply:
Otherwise, your comment sounds valid. I'm still yet to fully grok this advanced consensus discussion, though ... ;)

Reply:
@carllin Hehe, quite dense discussion. My head is spinning hot!

Well, your thinking is correct under the assumption that _tower always has root unless something is corrupted_. But, I think otherwise: _restored tower can legitimately have **no root** if the tower is very new and warming up_. I encoded the case as code and test respectively: https://github.com/solana-labs/solana/pull/10718/files#r470502855 and https://github.com/solana-labs/solana/pull/10718/files#r470501988

So, borrowing your logic, here is quite corner of corner of **Case 1** (not your **Case 2**) I've tried to address with the current code xD:

if ...
- (i) the newly-born tower just unfortunately starts to vote on a minor fork for several slots,
- (ii) and the validator is immediately stopped before rooting the tower,
- (iii) and the validator is later restarted with a snapshot which is **after** (not **before**) the common ancestor of both the minor fork and the major (of snapshot) fork (this should create a situation where _there are votes which aren't descendents of local root after replay_),

,then

- (x) the voted slots should be unavailable in the bank_forks and therefore,
- (y) there need to be some way (= stray slots) to exclude `lockouts` on an ancestors of the last vote because other validator might have legitimately voted on the same minor fork as well. Note that those ancestors can be newer than the local root (= `lockout_interval_start > root`).

Am I right? Be aware of that I'm still not expert in this consensus area. Or the your mentionedloptimistic confirmation rule doesn't need to enforced in this case?  I might be blatantly wrong... ;)

Reply:
@ryoqun my head is spinning too :)

Let me see if I can properly illustrate the example you mentioned above:

```
      1----------8---------10 (last vote on minor fork)
   /
0
    \
       2 (snapshot) -- 3
```

1) The validator voted on `1,  8, 10`, didn't set a root, then crashed
2) The validator restarts from a snapshot of `2`

Now let's consider your concern: "there need to be some way (= stray slots) to exclude lockouts on an ancestors of the last vote because other validator might have legitimately voted on the same minor fork as well. Note that those ancestors can be newer than the local root". 

So in our example:

1) "those ancestors can be newer than the local root" - The "local root" here would be slot `2`, so an example off such an ancestor would be slot `8 > 2` in the above example.

2) "there need to be some way (= stray slots) to exclude lockouts on an ancestors of the last vote". Here, because slot `8` is not descended from the root `slot 2`, it won't exist in BankForks, so it's naturally excluded.

Does that make sense? Or did I misunderstand your concern 😄

Reply:
Yeah, I've finally understood. It looks like we can ditch the stray restored ancestors! Thanks for explaining!

I'll promptly work on that.

Reply:
Ok, I've successfully reverted and ci is green: https://github.com/solana-labs/solana/pull/10718/commits/4fdb41ffd500a9f2894988869a843a09d2d1ee0c

I made the error handling a hard error, considering now I'm informed that these conditions are really odd.